src/main/java/org/springframework/data/convert/CustomConversions.java

```diff
@@ -0,0 +1,489 @@
+						.filter(it -> requestTargetTypeIsAssignable(targetType, it))//
+						.findFirst());
+	}
+
+	private static boolean hasAssignableSourceType(ConvertiblePair pair, Class<?> sourceType) {
+		return pair.getSourceType().isAssignableFrom(sourceType);
+	}
+
+	private static boolean requestTargetTypeIsAssignable(Optional<Class<?>> requestedTargetType, Class<?> targetType) {
+
+		return !requestedTargetType.isPresent() //
+				? true //
+				: requestedTargetType.map(it -> targetType.isAssignableFrom(it)).orElse(false);
+	}
+
+	/**
+	 * Conversion registration information.
+	 * 
+	 * @author Oliver Gierke
+	 * @author Mark Paluch
+	 */
+	@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+	private static class ConverterRegistration {
+
+		private final @NonNull ConvertiblePair convertiblePair;
+		private final @NonNull StoreConversions storeConversions;
+		private final boolean reading;
+		private final boolean writing;
+
+		/**
+		 * Returns whether the converter shall be used for writing.
+		 * 
+		 * @return
+		 */
+		public boolean isWriting() {
+			return writing == true || (!reading && isSimpleTargetType());
+		}
+
+		/**
+		 * Returns whether the converter shall be used for reading.
+		 * 
+		 * @return
+		 */
+		public boolean isReading() {
+			return reading == true || (!writing && isSimpleSourceType());
+		}
+
+		/**
+		 * Returns the actual conversion pair.
+		 * 
+		 * @return
+		 */
+		public ConvertiblePair getConvertiblePair() {
+			return convertiblePair;
+		}
+
+		/**
+		 * Returns whether the source type is a Mongo simple one.
+		 * 
+		 * @return
+		 */
+		public boolean isSimpleSourceType() {
+			return storeConversions.isStoreSimpleType(convertiblePair.getSourceType());
+		}
+
+		/**
+		 * Returns whether the target type is a Mongo simple one.
+		 * 
+		 * @return
+		 */
+		public boolean isSimpleTargetType() {
+			return storeConversions.isStoreSimpleType(convertiblePair.getTargetType());
+		}
+	}
+
+	/**
+	 * Value type to capture store-specific extensions to the {@link CustomConversions}. Allows to forward store specific
+	 * default conversions and a set of types that are supposed to be considered simple.
+	 *
+	 * @author Oliver Gierke
+	 */
+	@Value
+	@Getter(AccessLevel.PACKAGE)
+	@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+	public static class StoreConversions {
+
+		public static final StoreConversions NONE = StoreConversions.of(SimpleTypeHolder.DEFAULT, Collections.emptyList());
+
+		SimpleTypeHolder storeTypeHolder;
+		List<Object> storeConverters;
+
+		/**
+		 * Creates a new {@link StoreConversions} for the given store-specific {@link SimpleTypeHolder} and the given
+		 * converters.
+		 * 
+		 * @param storeTypeHolder must not be {@literal null}.
+		 * @param converters must not be {@literal null}.
+		 * @return
+		 */
+		public static StoreConversions of(SimpleTypeHolder storeTypeHolder, Object... converters) {
+
+			Assert.notNull(storeTypeHolder, "SimpleTypeHolder must not be null!");
```

Comment:
How so? You'd explicitly need to call `StoreConversions.of(new Object[] { collection })` to run into this problem, right?